conjure-java-core/src/main/java/com/palantir/conjure/java/types/MethodSpecs.java

```diff
@@ -84,42 +86,71 @@ public static MethodSpec createHashCode(Collection<FieldSpec> fields) {
                 // Single volatile read
-                .addStatement("int result = memoizedHashCode")
+                .addStatement("int result = $N", field)
                 .beginControlFlow("if (result == 0)")
-                .addStatement("result = $L", computeHashCode(fields))
+                .addCode(generateHashCode("result =", fields))
                 // Single volatile write
-                .addStatement("memoizedHashCode = result")
+                .addStatement("$N = result", field)
                 .endControlFlow()
                 .addStatement("return result")
-                .build());
+                .build();
     }
 
-    private static CodeBlock computeHashCode(Collection<FieldSpec> fields) {
-        if (fields.size() == 1) {
-            FieldSpec fieldSpec = Iterables.getOnlyElement(fields);
-            if (fieldSpec.type.isPrimitive()) {
-                return CodeBlock.of("$1T.$2N($3L)", fieldSpec.type.box(), "hashCode", createHashInput(fieldSpec));
+    @VisibleForTesting
+    static CodeBlock generateHashCode(String prefix, Collection<FieldSpec> fields) {
+        String varName = "hash";
+        CodeBlock.Builder builder = CodeBlock.builder();
+        List<CodeBlock> codeBlocks = computeHashCode(fields);
+        if (codeBlocks.size() == 1) {
+            builder.addStatement("$L $L", prefix, Iterables.getOnlyElement(codeBlocks));
+        } else {
+            builder.addStatement("int $N = 4441", varName);
+            for (CodeBlock codeBlock : codeBlocks) {
+                builder.addStatement(codeBlock);
             }
-            return CodeBlock.of("$1T.$2N($3L)", Objects.class, "hashCode", createHashInput(fieldSpec));
+            builder.addStatement("$L $N", prefix, varName);
+        }
+        return builder.build();
+    }
+
+    @VisibleForTesting
+    static List<CodeBlock> computeHashCode(Collection<FieldSpec> fields) {
+        if (fields.size() == 1) {
+            return ImmutableList.of(computeHashCode(Iterables.getOnlyElement(fields)));
         }
-        return CodeBlock.of("$1T.$2N($3L)", Objects.class, "hash", getHashInput(fields));
+
+        String varName = "hash";
+        ImmutableList.Builder<CodeBlock> builder = ImmutableList.builder();
+        for (FieldSpec field : fields) {
+            builder.add(CodeBlock.of("$N += ($N << 5) + $L", varName, varName, computeHashCode(field)));
+        }
+        return builder.build();
     }
 
-    private static CodeBlock getHashInput(Collection<FieldSpec> fields) {
-        return CodeBlocks.of(fields.stream().map(MethodSpecs::createHashInput).collect(joining(CodeBlock.of(", "))));
+    @VisibleForTesting
+    static CodeBlock computeHashCode(FieldSpec fieldSpec) {
+        if (fieldSpec.type.isPrimitive()) {
+            return CodeBlock.of("$1T.$2N($3L)", fieldSpec.type.box(), "hashCode", createHashInput(fieldSpec));
+        }
+        return CodeBlock.of("$1T.$2N($3L)", Objects.class, "hashCode", createHashInput(fieldSpec));
```

Comment:
Given we don't support nulls, we could do something like this:
```suggestion
        return CodeBlock.of("$L.hashCode()", createHashInput(fieldSpec));
```

Reply:
updated